presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPlanFunctions.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    @Description("Get the plan ids of given plan node")
+    @ScalarFunction("json_plan_node_children")
+    @SqlType("ARRAY<VARCHAR>")
+    @SqlNullable
+    public static Block jsonPlanNodeChildren(@TypeParameter("ARRAY<VARCHAR>") ArrayType arrayType,
+            @SqlType(StandardTypes.JSON) Slice jsonPlan,
+            @SqlType(StandardTypes.VARCHAR) Slice planId)
+    {
+        final List<JsonRenderedNode> planFragments = parseJsonPlanFragmentsAsList(jsonPlan);
+
+        if (planFragments == null) {
+            return null;
+        }
+
+        final Map<String, List<String>> planMap = extractPlanIds(planFragments);
+        List<String> planChildren = planMap.get(planId.toStringUtf8());
+        if (planChildren == null) {
+            return null;
+        }
+        return constructSqlArray(arrayType, planChildren);
+    }
+
+    @Description("Scrub the query plan")
+    @ScalarFunction("json_plan_scrub")
+    @SqlType(StandardTypes.JSON)
+    @SqlNullable
+    public static Slice jsonScrubPlan(@SqlType(StandardTypes.JSON) Slice jsonPlan)
+    {
+        Map<PlanFragmentId, Map<String, JsonRenderedNode>> jsonRenderedNodeMap = parseJsonPlanFragments(jsonPlan);
+
+        if (jsonRenderedNodeMap == null) {
+            return null;
+        }
+
+        jsonRenderedNodeMap.forEach((key, planMap) ->
+        {
+            planMap.put("plan", scrubJsonPlan(planMap.get("plan")));
+        });
+
+        final JsonCodec<Map<PlanFragmentId, Map<String, JsonRenderer.JsonRenderedNode>>> planMapCodec = constructJsonPlanMapCodec();
+        return utf8Slice(planMapCodec.toJson(jsonRenderedNodeMap));
+    }
+
+    private static Map<String, List<String>> extractPlanIds(List<JsonRenderedNode> jsonPlanFragments)
+    {
+        Map<String, List<String>> planMap = new HashMap<>();
+        jsonPlanFragments.stream().forEach(jsonPlanFragment -> addChildrenPlanIds(jsonPlanFragment, planMap));
+        return planMap;
+    }
+
+    private static JsonRenderedNode parseJsonPlan(Slice jsonPlan)
+    {
+        final JsonCodec<JsonRenderedNode> planCodec = constructJsonCodec();
+        try {
+            JsonRenderedNode jsonPlanFragment = planCodec.fromJson(jsonPlan.getBytes());
+            return jsonPlanFragment;
+        }
+        catch (Exception ex) {
+            // error parsing plan
+            return null;
+        }
+    }
+
+    private static List<JsonRenderedNode> parseJsonPlanFragmentsAsList(Slice jsonPlan)
+    {
+        Map<PlanFragmentId, Map<String, JsonRenderedNode>> jsonRenderedNodeMap = parseJsonPlanFragments(jsonPlan);
+
+        if (jsonRenderedNodeMap == null) {
+            return null;
+        }
+        List<JsonRenderedNode> planFragments = new ArrayList<>();
+        jsonRenderedNodeMap.values().forEach(map -> planFragments.addAll(map.values()));
+        return planFragments;
+    }
+
+    private static JsonCodec<Map<PlanFragmentId, Map<String, JsonRenderer.JsonRenderedNode>>> constructJsonPlanMapCodec()
```

Comment:
Is this using the same API used in plan serialization for logging? It will be better for consistency since I assume the input to these functions are plan json from logging.

Reply:
it uses the same basic class (JsonRenderedNode) but in the case of explain and logging the serialization happens directly from the plan node representation (to which we don't have access here)